File: assets/LF2/ServiceRequestLambda.ts
import {Handler, SQSEvent} from "aws-lambda";
import {SQS, DynamoDB} from "aws-sdk";
import {Client} from "@elastic/elasticsearch";
import {SendEmailCommand, SESClient} from "@aws-sdk/client-ses";

const dynamoDbClient = new DynamoDB();
const elasticSearchClient = new Client({
    node: 'https://search-restaurants-noe3v4hr6z2haev5f4tz4bsll4.us-east-1.es.amazonaws.com/',
});
const sqsClient = new SQS();
const sesClient = new SESClient();

async function getRestaurantId(cuisine: string) {
    const query = {
        index: 'kibana_1',
        body: {
            query: {
                match: {
                    Cuisine: 'chinese'
                }
            }
        }
    };

    let restaurantId: string | null = null;
    try {
        const response = await elasticSearchClient.search(query);

        // @ts-ignore
        console.log('Response: ', response.hits.hits[0]._source.name);

        const hitRandomIndex = Math.floor(Math.random() * response.hits.hits.length);

        // @ts-ignore
        restaurantId = response.hits.hits[hitRandomIndex]._source.id! as string;
    } catch (error) {
        console.error('Error: ', error);
        throw error;
    }

    return restaurantId;
}

async function getRestaurantInformation(cuisine: string, id: string) {
    const params = {
        Key: {
            'cuisine': {
                S: cuisine
            },
            'id': {
                S: id
            }
        },
        TableName: 'Restaurants'
    };

    const response = await dynamoDbClient.getItem(params).promise();
    return response.Item;
}

async function popSqsQueue(receiptHandle: string) {
    try {
        const response = await sqsClient.deleteMessage({
            QueueUrl: 'https://sqs.us-east-1.amazonaws.com/132900788542/SuggestionRequestQueue',
            ReceiptHandle: receiptHandle
        }).promise();
    } catch (error) {
        console.log('Error when deleting from queue: ', error);
    }
}

async function sendReservationEmail() {
    const sendEmailRequest = {
        Source: 'some source',
        Destination: {
            ToAddresses: [
                'email'
            ]
        },
        Message: {
            Subject: {
                Data: `Your reservation for DATE at TIME`
            },
            Body: {
                Text: {
                    Data: 'Reservation Details'
                }
            }
        }
    };

    const command = new SendEmailCommand(sendEmailRequest);

    try {
        const response = await sesClient.send(command);
    } catch (error) {
        console.error('Error: ', error);
    }
}

export const handler: Handler = async function(event: SQSEvent): Promise<void> {
    for (const record of event.Records) {
        try {
            const restaurantId = await getRestaurantId(record.messageAttributes.Cuisine.stringValue!);

            const restaurantInformation = await getRestaurantInformation(
                record.messageAttributes.Cuisine.stringValue!,
                restaurantId
            );

            console.log(restaurantInformation);

            await popSqsQueue(record.receiptHandle);
        } catch (error) {
            console.log('Retrying');
        }
    }
}

File: assets/LF2/ServiceRequestLambda.py
import boto3
import random

from boto3.dynamodb.conditions import Key
from elasticsearch import Elasticsearch, RequestsHttpConnection
from aws_requests_auth.aws_auth import AWSRequestsAuth
from botocore.exceptions import ClientError

sqs_client = boto3.client('sqs')
ses_client = boto3.client('ses')
dynamodb_client = boto3.client('dynamodb')

queue_url = 'https://sqs.us-east-1.amazonaws.com/132900788542/SuggestionRequestQueue'
elasticsearch_endpoint = 'search-restaurants-manual-ubb7tnfrs5osv6tew7ucbgsuxi.us-east-1.es.amazonaws.com'

request_authorization = AWSRequestsAuth(
    aws_access_key='AKIAR54MBDE7AZ7OGYE3',
    aws_secret_access_key='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',
    aws_region='us-east-1',
    aws_service='es',
    aws_host=elasticsearch_endpoint
)

elasticsearch_client = Elasticsearch(
    hosts=[{
        'host': elasticsearch_endpoint,
        'port': 443
    }],
    http_auth=request_authorization,
    use_ssl=True,
    verify_certs=True,
    connection_class=RequestsHttpConnection
)

def handler(event, context):
    service_request_dict = sqs_client.receive_message(
        QueueUrl=queue_url,
        MessageAttributeNames=['All'],
        AttributeNames=['All'],
        VisibilityTimeout=0,
        WaitTimeSeconds=0
    )

    service_request = service_request_dict['Messages'][0]
    retry_count = 3
    retry_flag = False

    try:
#         cuisine = service_request['MessageAttributes'].get('Cuisine').get('StringValue').lower()
        cuisine = 'Cuban'
        possible_restaurants = elasticsearch_client.search(
            index='data',
            body={
                'query': {
                    'match': {
                        'cuisine': cuisine
                    }
                }
            }
        )

        restaurant_id = random.choice(possible_restaurants['hits']['hits'])['_source'].get('id')
        restaurant_information = dynamodb_client.get_item(
            TableName='Restaurants',
            Key={
                'cuisine': {
                    'S': cuisine
                },
                'id': {
                    'S': restaurant_id
                }
            }
        )
        # Get attributes: restaurant_information['Items']['Name']
        send_email(
            service_request['MessageAttributes'].get('EmailAddress').get('StringValue'),
            'Restaurant recommendation'
        )

    except ClientError as e:
        print(f'Error: {e}')

    finally:
        sqs_client.delete_message(
            QueueUrl=queue_url,
            ReceiptHandle=service_request['ReceiptHandle']
        )

def send_email(email_address, email_body):
    try:
        response = ses_client.send_email(
            Destination={
                'ToAddresses': [
                    email_address
                ]
            },
            Message={
                'Body': {
                    'Text': {
                        'Data': email_body
                    }
                },
                'Subject': {
                    'Data': 'Restaurant Suggestion'
                }
            },
            Source='ishantaldekar1@gmail.com'
        )
        print('Email send')
    except ClientError as e:
        print(e.response['Error']['Message'])
